evennia/contrib/actions/handlers.py

```diff
@@ -0,0 +1,752 @@
+
+        Note: In turn-based mode, popping must only occur while the popped
+        action's owner is referenced by the actions_turnof variable
+        """
+        self.remove(action)
+        action = dict(action)
+        self.owner.location.actions.add(action)
+
+
+    def remove(self, action):
+        """
+        remove the action from the owner's actions queue
+        """
+        self.owner.msg("|mremoving {0} from action queue|n".format(action['key']))
+        self.list.remove(action)
+ 
+
+    def done(self):
+        """
+        During turn-based mode, concludes the turn for the owner
+
+        [WIP] Prohibit "done" from working when the character has issued a
+        movement action, crafting action or any other action with a sufficiently
+        large duration this turn, in order to prevent abuse of turn-based mode
+        for performing large numbers of commands very quickly.
+        """
+        mode = self.owner.location.actions.mode
+        turnof = self.owner.location.actions.turnof
+        if mode == "TB" and turnof == self.owner:
+            # Bring the action system forward
+            self.owner.location.actions.run_script()
+        elif mode == "TB":
+            self.owner.msg("You cannot finish your turn without it being " +
+                "your turn to begin with.")
+        else:
+            self.owner.msg("You cannot finish your turn without being in " +
+                "turn-based mode to begin with.")
+
+    def stop(self, ongoing=True, queued=False):
+        """
+        Cancels the character's ongoing actions if ongoing is set to True,
+        and removes the character's queued actions if queued is set to True.
+        """
+        if queued:
+            # First stop all enqueued actions if the action calls for it, 
+            # to ensure that cancelling the ongoing actions will not load
+            # any actions from the queue
+            for action in self.owner.actions.list:
+                self.owner.actions.remove(action)
+
+        if ongoing:
+            # Then cancel all ongoing actions
+            for action in self.owner.location.actions.list:
+                if action['owner'] == self.owner:
+                    action['at_cancel'](action)
+
+
+    def unpuppet(self):
+        """
+        Cancels all of the character's queued and ongoing actions, then
+        removes the character's alert, which causes the room to attempt to
+        enter real-time mode.
+        """
+        self.stop(ongoing=True, queued=True)
+        self.turnbased = False
+
+
+class RoomActionHandler(object):
+    """
+    Lists the ongoing action in the room, keeps track of turn-related
+    information and provides methods for 
+    """
+
+    def __init__(self, owner):
+        """
+        sets owner (Room), the room to which this handler is attached
+        """
+        self.owner = owner
+
+
+    def setup(self, override=False):
```

Comment:
This requires a docstring detailing when this is called and maybe why it's here alongside `__init__`.